### DISCUSSION TITLE
Support toggling editing of all selections versus only primary selection

### DISCUSSION THREAD

**Discussion by EpocSquadron** [helix-editor/helix#10457] [Ideas]
One pain point I've had as a kakoune user with regards to multiple cursors is a desire to occasionally modify only one of a set of selections. That is, of like to be able to suspend editing of all but the primary selection, make a modification, then resume editing all of the selections without losing my place.

## Motivating Example

I have a list of words that I would like to turn into elements of an array in a JSON document.

```
Some set of words
```

I will first select the whole line `x`, surround it with square brackets `ms[`, then split on space `S\s+<enter>`. Now I have a selection for each word, and can further interact to put them on their own lines `i<enter><esc>`, and quote them `ms"`, and finally place a comma after each `a,<esc>`, arriving at the following:

```json
[
  "Some",
  "set",
  "of",
  "words",]
```

This isn't yet valid as the trailing comma on the final array item is not allowed. I could collapse to the primary selection and move over to the comma, replace it with a newline `c<enter><esc>`, but now I've lost the selections I had previously. In this example that doesn't seem so bad but I've come across situations where I want to correct the last item, then continue to act on the set of selections. Often the next operation might be a regex replacement for which I need to have consistent pattern, where the ending brace might cause a problem. Or it might be the opposite of the above example and I need to anchor on the trailing comma but it isn't present on the last selection due to having been properly formatted previously. In any case I often wish to be able to suspend editing of the non-primary selections, correct that one item that breaks the mold, then resume editing all selections.

## Proposed Interface

Key bindings (up for discussion on better ones)

* `Alt-,` - when pressed in normal mode, toggle ability to modify all selections. There is a danger here of if you accidentally miss the alt modifier, you lose all but your primary, so it might be better to swap it so that `,` is the toggle and `alt-,` discards. But then discarding other selections becomes less ergonomic. 
* `(`, `)` - cycle primary selection to next or previous selection. This is an existing keybind, but would need modification to retain (or just not alter) the state of the above toggle. These are significant because of you are in edit-primary state, you can cycle through the list naturally to make unique edits to every item, before eventually returning to edit-all state.

There should be visual feedback of which editing state we are in, by specifying another syntax color for selections that are not being edited. Ideally there would be three color grades:

* Primary selection - should have the brightest background color
* Secondary selection (being edited) - muted background color. I think this doesn't happen right now, instead using the cursor color to distinguish, but I think this is a little more easily understood.
* Secondary selection (not being edited) - very muted or even grey or some other color.

Obviously this would be up to the color scheme, but we could establish this add an expected pattern to implement in any new ones and revisit the existing.

We may also want to throw a state indicator in the diagnostic line.

## Alternatives

One alternative to implementing this feature is to encourage the user to save their selections, collapse to the primary, edit that selection, then restore the previous selections. This was what I did in kakoune using the "marks" feature. To my knowledge, an equivalent feature is not implemented in helix yet. Further, even if it were I would argue that there is still a usability case for this feature over telling people to use marks.

* I think it might be more natural for some people to wonder how they can do what is requested rather than wonder how they can save and return to selections.
* Marks, when present, might be being used to save another important selection set, forcing the user to use a different register to save and restore the selections, which can be more cumbersome than what is requested here.

Some evidence for the desire for a feature like this exists in the form of a plugin for kakoune called [Phantom selections](https://github.com/occivink/kakoune-phantom-selection), which incidentally users marks internally.

**Comment by archseer** [comment]
We don't support this yet, but kakoune has `C-s` which pushes the current set of selections on the jumplist. You could `C-s`, then reduce to a single selection, do your edits then `C-o` to restore the selection set. It looks similar to your marks suggestion.

**Comment by EpocSquadron** [comment]
I had forgotten about `ctrl-s`. But yeah, I still think there's a case for this requested feature instead. If I were to work on this feature would it be welcome?

**Comment by Omnikar** [comment]
Adding this would also open up the possibility of manually placing cursors without the use of the mouse by turning off multi-cursor control, moving the primary cursor to a location, then pressing `C` to add an additional cursor there. Also, we might want to add something to the `_ sels` part of the statusline to indicate when multi-cursor control is disabled.

**Comment by EpocSquadron** [comment]
@Omnikar that's a brilliant foresight! Thinking about the case when you only have one selection and toggle this feature on, how should this behave? On one hand, it has no meaning until you use C to get a new selection. On the other, if we freeze the current selection and immediately spawn a new one on the first movement, we're acting differently than than in the multiple selection case. I'm thinking keep it simple and do the former. 

Another interesting downstream effect is the meaning of C. If you are using this feature to add cursors, you might expect to make the selection you want then press something to add it to the inactive selections and move on. With C and alt-C, you going the selection above or below instead. Perhaps a ctrl-c binding (do we use this to cancel pipe shells?) to add a new selection to the right of the current selection that is one character wide, and make it the primary pushing what was the primary selection into the list of secondaries. That could generalize to multiple cursors in the edit-all state as well, although it's utility would be considerably more niche, enough to consider only having it in this new edit-primary state.

**Comment by EpocSquadron** [comment]
Another thought, I believe that most editing primitives assume non-overlapping ranges, and will combine/flatten them if they do. With this, we would need to avoid flattening selections until we toggle back to edit-all state, or push a new non-primary selection (with `C`). That would allow navigating through an existing selection without inadvertently erasing it, while still preventing awkward overlapping selections from being created and causing havoc.

**Comment by aral** [comment]
Interesting, when I follow the instructions in your issue, I get:

```
[
    "S"ome 
    "s"et 
    "o"f 
    "w"ords]
```

(After entering insert mode and pressing enter, the selection seems to revert to just the first letter.)

helix 22.03 (c18de0e8) on elementary OS 6.

**Comment by the-mikedavis** [comment]
The `ms"` step surrounds the selection and it looks like the <kbd>i</kbd><kbd>Enter</kbd> collapses the selctions to the first character. So it'd need an extra `w` (or I suppose `miw` to be more generic) after the <kbd>i</kbd><kbd>Enter</kbd>.

**Comment by aral** [comment]
@the-mikedavis Do you know if that’s as designed or a regression? I find it surprising that inserting collapses the selections. I would expect them to remain as per the original instructions.

**Comment by the-mikedavis** [comment]
I'll have to do a bisect to be sure. It seems like the selections should remain when inserting newlines though.
